refactor(recs): replace `EntityIndex/EntityId` with `EntitySymbol/Entity` and expose only `Entity`

alvrs opened the pull request:
* Every component stores a mapping between entity and its value in the component. For memory efficiency we're not using the full 32 byte hex string as index but locally map it to a number and use the number as index. 
* This caused some confusion for consumers: methods like `getComponentValue` expected `EntityIndex` (the number) as input, but when calling contracts the `EntityID` (the hex string) was expected
* This PR changes this to only expose the `Entity` (the hex string) to consumers, and internally maps it to a javascript native Symbol for memory efficiency, which is hidden from consumers

BREAKING CHANGE:
- Previously consumers had to think about when to use `EntityIndex` and when to use `EntityID`. With this change, only a single entity type (`Entity`) is exposed to consumers. To upgrade, replace all your references to `EntityIndex` with `Entity` (including any logic that expects this value to be a number / any logic that expect the need to do manual conversion between `EntityID` and `EntityIndex`)
- Previously the `recs` world exposed a `entities` array. This is replaced by a `getEntities()` method returning an Iterable of Entities. (This is to abstract internally used `EntitySymbol` away from consumers)

holic commented on packages/recs/src/types.ts:
```diff
@@ -1,16 +1,14 @@
- * Used to refer to the string id of an entity (independent from a {@link World}).
- */
-export type EntityID = Opaque<string, "EntityID">;
+export type Entity = Opaque<string, "Entity">;
```
is there an advantage of using this custom type instead of just a string? are plain strings compatible with this type?

holic commented on packages/std-client/src/setup/setupMUDV2Network.ts:
```diff
@@ -178,6 +178,7 @@ export async function setupMUDV2Network<C extends ContractComponents>({
     components,
     singletonEntityId: SingletonID,
     singletonEntity,
+    /* @deprecated playerEntityId is equivalent to playerEntity */
```
should we guide folks on which one to use?

alvrs replied:
I'm not sure if the `deprecated` tag here shows up in vscode, but the intention was to deprecate `playerEntityId` and just use `playerEntity`

holic reviewed (approved):
so much easier than I expected! nice!

only maybe-blocking thing we should probably do here is prefix the symbol, since it's very global

alvrs commented on packages/recs/src/Component.ts:
```diff
@@ -49,7 +47,8 @@ export function defineComponent<S extends Schema, M extends Metadata, T = undefi
   const metadata = options?.metadata;
-  const entities = () => (Object.values(values)[0] as Map<EntityIndex, unknown>).keys();
+  const entities = () =>
+    transformIterator((Object.values(values)[0] as Map<EntitySymbol, unknown>).keys(), getEntityString);
```
yep exactly

alvrs commented on packages/recs/src/Entity.ts:
```diff
@@ -29,3 +29,18 @@ export function createEntity(
+ * Entities are represented as symbols internally for memory efficiency.
+ */
+export function getEntitySymbol(entityString: string): EntitySymbol {
+  return Symbol.for(entityString) as EntitySymbol;
```
we could, but do we need to? Even if something else outside of MUD is using the same symbol, I think that shouldn't be an issue. If we don't need to I'd like to avoid string manipulations here for performance reasons (eg. we apply `getEntityString` for every symbol when calling `world.getEntities()`, so any additional operation we do in there might cause issues if we're handling lots of entities)

holic replied:
ah good point, there isn't much of an issue with overlaps

alvrs commented on packages/ecs-browser/src/Browser.tsx:
```diff
@@ -23,20 +23,20 @@ export const Browser = observer(
         <QueryBuilder
           devHighlightComponent={devHighlightComponent}
-          allEntities={world.entities}
+          allEntities={[...world.getEntities()]}
```
Good point! Wonder if we should expose the array of entity symbols too for use cases like this

alvrs replied:
Exposed `entitySymbols`, but gonna leave `Browser.tsx` as is since it's deprecated and I wanna avoid deeper changes here

alvrs commented:
Only tiny change since last approval was exposing the `entitySymbols` set in case consumers need a stable reference: https://github.com/latticexyz/mud/pull/748/commits/3fd27bc26304c7930de6a4b8205194b1d3fad260

holic commented:
cc @nickcernera as this will impact v2 emojimon (in the form of simplifying a bunch of code)
